Review

endlesslydivided opened the pull request:
Review here!

endlesslydivided committed:
Entities classes are created, typeorm is configured, auth decorator, middleware and guard are created,; client routing, layout, redux store and api are created and configured

endlesslydivided committed:
Play session, quiz, answer, result and reporn entities, controllers and services are created. Playing session gateway is created and configured. Client: quizzes list and form are created, socket provider is created and configured, routing is configured, stores and api slices are created. Zoom context exchange is set up.

endlesslydivided committed:
Quiz answering enabled

endlesslydivided committed:
Quiz activation block enabled without meeting. Result list created. Front-end context provision enabled.

endlesslydivided committed:
Pagination is created. Results list is edited. Quiz delete is enabled. Server quiz soft delete is implemented. ZoomSDK is set up and token retrievement is implemented.

endlesslydivided committed:
Zoom api access token retrievemen, acess token and refresh token guard,middleware and decorators are created. User interface design is improved. Quiz play sessions reports are enabled.

endlesslydivided committed:
Screens and video are inclluded

endlesslydivided committed:
Create README.md

endlesslydivided committed:
Refactored

endlesslydivided committed:
Merge branch 'main' of https://github.com/endlesslydivided/ZOOM_QUIZ_APP into main

endlesslydivided committed:
Backend refactored: all explicit and implicit any mentions are removed, variables and function return types are writter. Auth service is edited. Modules are edited.

endlesslydivided committed:
Docker compose, dockerfiles are created. Play-sessions services tests are created.

endlesslydivided committed:
Play sessions controller tests are added. Quizzes controller  (unfinished) and service tests are added. Results service tests are added

endlesslydivided committed:
Auth service and controller test are finished. React refactored and types written. Docker compose is edited and postgres docker setup is created

MaksimLihoi commented on Client/src/App.scss:
```diff
@@ -0,0 +1,77 @@
+/*#region CSS overrides*/
+*
```
you use different code styles in ur scss files. keep only one and use prettier

MaksimLihoi commented on Client/src/App.tsx:
```diff
@@ -1,14 +1,57 @@
-import React from 'react';
-import './App.css';
+import zoomSdk from "@zoom/appssdk";
```
use prettier

MaksimLihoi commented on Client/src/App.tsx:
```diff
@@ -1,14 +1,57 @@
+    const configAuth = async () => 
+    {
+    
+      const result = await zoomSdk.config({capabilities: ["getRunningContext","getSupportedJsApis","openUrl","authorize","onAuthorized"]})
```
from my point of view it will be better to move these keys ("getRunningContext","getSupportedJsApis","openUrl","authorize","onAuthorized") to enum and use enum values instead of string

MaksimLihoi commented on Client/src/App.tsx:
```diff
@@ -1,14 +1,57 @@
+      {
+        code:event.code,
+        state:event.state,
+        verifier:'ac3e722ede6ff88fac10cb6e02e2f63ab0acf08148f91b699988e097',
```
what is it?

MaksimLihoi commented on Client/src/App.tsx:
```diff
@@ -1,14 +1,57 @@
+    zoomSdk.authorize({state,codeChallenge:challenge})
+    .catch((e) => 
+    {
+      notification.error({message:'Some erro occured during ZoomSDK authorize. You better restart the app.'});
```
such messages will be better to keep as enum

MaksimLihoi commented on Client/src/components/AppRouter/AppRouter.tsx:
```diff
@@ -0,0 +1,75 @@
+import './AppRouter.scss';
+
+
+interface AppRouterProps {
```
useless interface. why do we have to keep an empty interface?

MaksimLihoi commented on Client/src/components/AppRouter/AppRouter.tsx:
```diff
@@ -0,0 +1,75 @@
+import { Typography } from 'antd';
+import React, { useEffect } from 'react';
+import { Link, Navigate, Route, Routes } from 'react-router-dom';
+import { useAppDispatch, useAppSelector } from '../../hooks/redux';
```
also i suggest u to use alias

MaksimLihoi commented on Client/src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+import './Loader.scss'
+
+interface LoaderProps {
+    containerStyle?: object;
```
What specific information about "containerStyle" will the "object" type provide us with?

MaksimLihoi commented on Client/src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+
+interface LoaderProps {
+    containerStyle?: object;
+    spinProps?:object;
```
the same

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+const PlaySessionItem: React.FC<PlaySessionItemProps> = ({ playSession }) => {
+
+    const isResult = !!playSession.results && playSession.results?.length !== 0;
```
`Boolean(playSession.results)` will be better

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+    const isResult = !!playSession.results && playSession.results?.length !== 0;
+
+    const userAnswerId = isResult && playSession.results![0].answer?.id;
```
Why do use `playSession.results!`? i mean `!`

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+                <Collapse className="answers-collapse">
+                    <Collapse.Panel header="Reveal answers" key="1">
+                        {
+                            playSession.quiz.answers?.map((answer: any) => <PlaySessionAnswerItem answer={answer} playSession={playSession} />)
```
Do not use any type. Specify with particular type.

MaksimLihoi replied:
Also u forget about key prop

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+
+const PlaySessionAnswerItem = ({ answer, playSession }: { answer: Answer, playSession: UserPlaySession }) => {
+    const socket: any = useContext(SocketContext);
```
replace any

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+    const isUserAnswer = isResult && playSession.results![0].answer?.id === answer.id;
+
+    const isCorrectAnswer = isResult && playSession?.quiz?.answers?.find((a: any) => a.id === answer.id)?.isCorrect;
```
the same about any

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+    const userAnswerId = isResult && playSession.results![0].answer?.id;
+
+    const isCorrect = isResult && playSession?.quiz?.answers?.find((a: any) => a.id === userAnswerId)?.isCorrect;
```
replace any

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionItem.tsx:
```diff
@@ -0,0 +1,87 @@
+}
+
+
+const PlaySessionAnswerItem = ({ answer, playSession }: { answer: Answer, playSession: UserPlaySession }) => {
```
move this component to separated file

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionList.tsx:
```diff
@@ -0,0 +1,57 @@
+import './PlaySessionList.scss';
+import { UserPlaySession } from '../../types/storeSliceTypes';
+
+interface PlaySessionListProps {}
```
empty interface

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionList.tsx:
```diff
@@ -0,0 +1,57 @@
+
+    useNotify({
+        result: playSessions,
+        successCB: () => {
```
not clear name `successCB`

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionList.tsx:
```diff
@@ -0,0 +1,57 @@
+    });
+
+    return (
+        <Space size={'middle'} direction='vertical'>
```
`size={'middle'} direction={'vertical'}` 
or
`size='middle' direction='vertical'` 

but use prettier

MaksimLihoi commented on Client/src/components/PlaySessionList/PlaySessionList.tsx:
```diff
@@ -0,0 +1,57 @@
+                grid={{ gutter: 10, column: 1 }}
+                dataSource={playSessionsState}
+                renderItem={(playSession: UserPlaySession) => (
+                    <PlaySessionItem playSession={playSession} />
```
i guess u missed key prop here

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+    const [createQuiz] = useCreateQuizMutation();
+    const dispatch = useAppDispatch();
+    
+    const onFinish = async (values:any) => 
```
replace any

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+    
+    const onFinish = async (values:any) => 
+    {
+      const body:any = {text:values.text,answers:[]}; 
```
the same

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+
+      body.answers[values.correctNumber].isCorrect = true;
+
+      const {data,error}:any = await createQuiz({body});
```
the same

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+          </Form.Item>
+        </Col>
+        <Col  span={24}>
+          <Form.Item name="answers1" label="1." colon={false}  rules={[{ required: true, message: 'Missing answer text' }]}>
```
strange prop value `answers1`

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+          </Form.Item>
+        </Col>
+        <Col  span={24}>
+          <Form.Item name="answers2" label="2." colon={false}  rules={[{ required: true, message: 'Missing answer text' }]}>
```
the same

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+          </Form.Item>     
+        </Col>
+        <Col  span={24}>
+          <Form.Item name="answers3" label="3." colon={false}  rules={[{ required: true, message: 'Missing answer text' }]}>
```
the same

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+          </Form.Item>    
+        </Col>
+        <Col  span={24}>
+          <Form.Item name="answers4" label="4." colon={false}  rules={[{ required: true, message: 'Missing answer text' }]}>
```
the same

MaksimLihoi commented on Client/src/components/QuizForm/QuizForm.tsx:
```diff
@@ -0,0 +1,111 @@
+        <Col  span={24}>
+          <Form.Item label="Correct answer: " name="correctNumber" colon={false}  rules={[{ required: true, message: 'Missing correct answer' }]}>
+            <Radio.Group>
+              <Radio value={0}>1.</Radio>
```
looks like code duplicate

MaksimLihoi replied:
maybe use map instead

MaksimLihoi commented on Client/src/components/QuizList/QuizList.tsx:
```diff
@@ -0,0 +1,65 @@
+import { UserQuiz } from "../../types/storeSliceTypes";
+
+interface QuizListProps
+{
```
the same

MaksimLihoi commented on Client/src/components/QuizList/QuizList.tsx:
```diff
@@ -0,0 +1,65 @@
+        }
+    ) 
+
+    const onChangePage: PaginationProps['onChange'] = (page) => setFilters((p:any) => {return {...p, page}});
```
setFilters(p => ({...p, page}));

MaksimLihoi commented on Client/src/components/QuizList/QuizList.tsx:
```diff
@@ -0,0 +1,65 @@
+                size={"small"} 
+                dataSource={quizzes}
+                renderItem={(quiz:UserQuiz) => <QuizListItem quiz={quiz}/>}/>
+            <Pagination current={filters.page} defaultPageSize={filters.limit} total={pagesCount} onChange={onChangePage}/>
```
Instead of defining two identical Pagination components, you can create a separate Paginator component and reuse it:
```
const Paginator = () => (
  <Pagination current={filters.page} defaultPageSize={filters.limit} total={pagesCount} onChange={onChangePage} />
);

// Usage
<>
  <Paginator />
  <List ... />
  <Paginator />
</>

```

MaksimLihoi commented on Client/src/components/QuizList/QuizList.tsx:
```diff
@@ -0,0 +1,65 @@
+                grid={{gutter:10,column:1}}
+                size={"small"} 
+                dataSource={quizzes}
+                renderItem={(quiz:UserQuiz) => <QuizListItem quiz={quiz}/>}/>
```
maybe key prop

MaksimLihoi commented on Client/src/components/QuizList/QuizListItem.tsx:
```diff
@@ -0,0 +1,78 @@
+
+    const context = useAppSelector((state:RootState) => state.zoomContext.decrypted);
+    const {onDeleteClickHandler} = useDeleteQuiz({entity:quiz})
+    const socket:any = useContext(SocketContext);
```
any

MaksimLihoi commented on Client/src/components/QuizList/QuizListItem.tsx:
```diff
@@ -0,0 +1,78 @@
+                    <Collapse.Panel header="Reveal answer" key="1">
+                    {
+                        quiz.answers.map((answer:Answer) =>
+                        <Card className={`answer-text ${answer.isCorrect  ? 'correct-answer' : 'wrong-answer'}`}>
```
key prop

MaksimLihoi commented on Client/src/components/QuizList/QuizStatsModal.tsx:
```diff
@@ -0,0 +1,79 @@
+
+const QuizStatsModal:React.FC<QuizStatsModalProps> = ({isModalOpened,setIsModalOpened,playSessions}) => {
+
+    const [report,setReport]:any = useState(null);
```
any

MaksimLihoi commented on Client/src/components/QuizList/QuizStatsModal.tsx:
```diff
@@ -0,0 +1,79 @@
+    const [report,setReport]:any = useState(null);
+    const [getReport] = useLazyGetPlaySessionReportQuery();
+
+    const onChangeSelect = async (value:any) =>
```
any

MaksimLihoi commented on Client/src/components/QuizList/QuizStatsModal.tsx:
```diff
@@ -0,0 +1,79 @@
+
+    const onChangeSelect = async (value:any) =>
+    {
+        const {data} = await getReport({psId:value})
```
`psId` not clear name

MaksimLihoi commented on Client/src/components/QuizList/QuizStatsModal.tsx:
```diff
@@ -0,0 +1,79 @@
+            {
+                report ?
+                <>
+                    <Statistic valueStyle={{ color: 'rgba(9, 208, 59, 0.771)' }} title="Correct answers:" value={report?.correctAnswersCount} />
```
Instead of hardcoding the color values, consider defining them as constants or variables, so they can be easily changed if needed.

MaksimLihoi commented on Client/src/components/QuizList/QuizStatsModal.tsx:
```diff
@@ -0,0 +1,79 @@
+                report ?
+                <>
+                    <Statistic valueStyle={{ color: 'rgba(9, 208, 59, 0.771)' }} title="Correct answers:" value={report?.correctAnswersCount} />
+                    <Statistic valueStyle={{ color: 'rgba(214, 80, 80, 0.633)' }} title="Incorrect answers:" value={report?.incorrectAnswersCount} />
```
Instead of hardcoding the color values, consider defining them as constants or variables, so they can be easily changed if needed.

MaksimLihoi commented on Client/src/components/ResultsList/ResultsList.tsx:
```diff
@@ -0,0 +1,53 @@
+        }
+    ) 
+
+    const onChangePage: PaginationProps['onChange'] = (page) => setFilters((p:any) => {return {...p, page}});
```
the same as above

MaksimLihoi commented on Client/src/components/ResultsList/ResultsList.tsx:
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+interface ResultsListProps
+{
```
same

MaksimLihoi commented on Client/src/components/ResultsList/ResultsListItem.tsx:
```diff
@@ -0,0 +1,84 @@
+
+const ResultsItem:React.FC<ResultsItemProps>= ({result}) =>
+{
+    const isResult = !!result.results && result.results?.length !== 0;
```
same

MaksimLihoi commented on Client/src/components/ResultsList/ResultsListItem.tsx:
```diff
@@ -0,0 +1,84 @@
+
+    const userAnswerId = isResult && result.results![0].answer?.id;
+
+    const isCorrect = isResult && result?.quiz?.answers?.find((a:any) => a.id === userAnswerId)?.isCorrect;
```
same

MaksimLihoi replied:
not clear name `a`

MaksimLihoi commented on Client/src/components/ResultsList/ResultsListItem.tsx:
```diff
@@ -0,0 +1,84 @@
+
+    const isCorrect = isResult && result?.quiz?.answers?.find((a:any) => a.id === userAnswerId)?.isCorrect;
+    
+    const yourAnswerText = isResult ? isCorrect ? " Correct!" : " Incorrect(" : " No answer";
```
Ninja code, not readable

MaksimLihoi commented on Client/src/components/ResultsList/ResultsListItem.tsx:
```diff
@@ -0,0 +1,84 @@
+const ResultsAnswerItem = ({answer,result}:{answer:Answer,result:UserPlaySessionResult}) =>
+{
+
+    const isResult = !!result.results && result.results?.length !== 0;
```
same

MaksimLihoi commented on Client/src/components/ResultsList/ResultsListItem.tsx:
```diff
@@ -0,0 +1,84 @@
+
+    const isUserAnswer =isResult && result.results![0].answer?.id === answer.id;
+
+    const isCorrectAnswer = isResult && result?.quiz?.answers?.find((a:Answer) => a.id === answer.id)?.isCorrect;
```
same

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.scss:
```diff
@@ -0,0 +1,7 @@
+{
+    .ant-notification-notice-message
+    {
+        margin:0 !important;
```
are u sure about important ? it's not recommended to use it.

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+    children: string | JSX.Element | JSX.Element[]
+}
+
+export const SocketContext:any = createContext(null);
```
any

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+
+const io = require('socket.io-client');
+
+export enum QuizClientEvent 
```
maybe it will be better to move enums to separated file specified for enums

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+    SERVER_PROVIDES_QUIZ = 'SERVER_PROVIDES_QUIZ'
+}
+
+export enum QuizServerEvent 
```
same

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+
+    const [socket,setSocket]:any = useState(null);
+    const [api, contextHolder] = notification.useNotification();
+    const openNotification = ({message,description,key}:any) => {
```
any

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+
+    useEffect(() =>
+    {
+        if (!socket) 
```
will be better not to start conditions with `!`

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+                    key:data.id,
+                    message:(
+                        <Row gutter={[10,10]} style={{flexWrap:'nowrap',cursor:'pointer'}} onClick={() => {
+                            navigate(`${ANSWER_ROUTE}`)
```
?

MaksimLihoi commented on Client/src/components/SocketProvider/SocketProvider.tsx:
```diff
@@ -0,0 +1,115 @@
+            }
+          });
+
+          socket.on(QuizClientEvent.SERVER_SENDS_ANSWERS,async (data: any) => {
```
any

MaksimLihoi commented on Client/src/hooks/useDeleteQuiz.ts:
```diff
@@ -0,0 +1,35 @@
+
+interface UseDeleteQuizParams
+{
+    entity:any;
```
any

MaksimLihoi commented on Client/src/hooks/useDeleteQuiz.ts:
```diff
@@ -0,0 +1,35 @@
+
+    useNotify({
+        result:deleteQuizResult,
+        successCB:()=>  
```
same about name

MaksimLihoi commented on Client/src/hooks/useNotify.ts:
```diff
@@ -0,0 +1,22 @@
+        {   
+            const message = errorMessage || result.error?.data?.message;
+            notification.error({message,placement:"topRight",duration:2.5});
+            errorCB && errorCB();
```
not clear name

MaksimLihoi commented on Client/src/hooks/useNotify.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const useNotify = (
+    {result,successMessage,successCB,errorMessage,errorCB}:
+    {result: any,successMessage?:string,successCB?:(...args:any[]) => void,errorMessage?:string,errorCB?:(...args:any[]) => void}) => {
```
any

MaksimLihoi commented on Client/src/hooks/usePagination.ts:
```diff
@@ -0,0 +1,62 @@
+    page: 1,
+    limit: 5,
+};
+
```
It might be a good idea to add some type annotations to return value of the usePagination hook to improve the readability and the maintainability of the code.

MaksimLihoi commented on Client/src/index.tsx:
```diff
@@ -1,14 +1,50 @@
+                    <li/>
+                    <li/>
+                    <li/>
+                    <li/>
```
?

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+import './SmMdLayout.scss';
+import React from 'react';
+
+//@ts-expect-error
```
?

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+
+export default function SmMdLayout() 
+{
+    const zoomContext:any = useAppSelector((state:any) => state.zoomContext.context);
```
any

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+
+
+
+export default function SmMdLayout() 
```
not clear name

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+    const zoomContext:any = useAppSelector((state:any) => state.zoomContext.context);
+
+
+    const user = useAppSelector((state:any) => state.zoomContext?.user)
```
any

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+    const items: MenuProps['items'] = [
+        {
+            label: (<Typography.Text strong type='secondary'>
+                        <NavLink to={`../${QUIZZES_ROUTE}?context=${zoomContext}`}>Your quizzes</NavLink>
```
hardcode

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+        },
+        {
+            label: (<Typography.Text strong type='secondary'>
+                        <NavLink to={`../${ANSWER_ROUTE}?context=${zoomContext}`}>Answer!</NavLink>
```
hardcoded param

MaksimLihoi commented on Client/src/layouts/SmMdLayout/SmMdLayout.tsx:
```diff
@@ -0,0 +1,81 @@
+        },
+        {
+            label: (<Typography.Text strong type='secondary'>
+                        <NavLink to={`../${RESULTS_ROUTE}?context=${zoomContext}`}>Results</NavLink>
```
same

MaksimLihoi replied:
also code duplicate

MaksimLihoi commented on Client/src/services/ApiSlice.ts:
```diff
@@ -0,0 +1,70 @@
+});
+
+const baseQueryWithReauth = async (
+    args?: any,
```
any

MaksimLihoi commented on Client/src/store/slices/PlaySessionsSlice.ts:
```diff
@@ -0,0 +1,45 @@
+import { Slice, createSlice } from "@reduxjs/toolkit";
+import { UserPlaySession } from "../../types/storeSliceTypes";
+
```
fix all any types

MaksimLihoi commented on Client/src/store/slices/PlaySessionsSlice.ts:
```diff
@@ -0,0 +1,45 @@
+    setPlaySessions:(state:any,action:any) =>
+    {
+      const data:UserPlaySession[] = action.payload;
+      const retrievedSessions:UserPlaySession[] = data.filter((c:any) => !state?.some((i:any) => i.id === c.id));
```
`(c:any) => !state?.some((i:any) => i.id === c.id)` with such names and any type i don't understand anything

MaksimLihoi commented on Client/src/store/slices/PlaySessionsSlice.ts:
```diff
@@ -0,0 +1,45 @@
+      const data:UserPlaySession[] = action.payload;
+      const retrievedSessions:UserPlaySession[] = data.filter((c:any) => !state?.some((i:any) => i.id === c.id));
+
+      const editedSessions:UserPlaySession[] = retrievedSessions.filter((r:any) => state?.some((c:any) => c.id === r.id));
```
same

MaksimLihoi commented on Client/src/store/slices/PlaySessionsSlice.ts:
```diff
@@ -0,0 +1,45 @@
+
+      const newSessions:UserPlaySession[] = retrievedSessions.filter((c:any) => !state?.some((r:any) => c.id === r.id));
+
+      const appliedSessions:UserPlaySession[] = [...state!.map((c:any) => editedSessions.map((e:any) => c.id === e.id ? e : c)[0] || c)];
```
not clear

MaksimLihoi commented on Client/src/utils/consts.ts:
```diff
@@ -0,0 +1,3 @@
+export const QUIZZES_ROUTE = '/'
```
it will be better to keep routes as enums

MaksimLihoi commented:
Nice code in general despite few little issues! 
Necessarily use prettier!

MaksimLihoi commented:
Also keep in mind rules of branch naming!

endlesslydivided committed:
Refactored: types written, project structure edited, consts grouped

endlesslydivided committed:
DB logic from services is moved to repositories. Tests are edited. Linter import sort options is added

endlesslydivided committed:
1. Client project structured modified.
2. TypeOrm migration set up, databases module created.
3. ConfigModule enabled.
4. Pre-commit eslint hook created.
5. .env removed and .env.sample created.

endlesslydivided committed:
1. Client project structured modified.
2. TypeOrm migration set up, databases module created.
3. ConfigModule enabled.
4. Pre-commit eslint hook created.
 5. .env removed and .env.sample created.
